common/blobstore/filestore/util.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package filestore
+
+import (
+	"bytes"
+	"encoding/gob"
+	"errors"
+	"github.com/uber/cadence/common/blobstore"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"os"
+)
+
+const (
+	dirMode  = os.FileMode(0700)
+	fileMode = os.FileMode(0600)
+)
+
+func fileExists(filepath string) (bool, error) {
+	info, err := os.Stat(filepath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	if info.IsDir() {
+		return false, errors.New("specified directory not file")
+	}
+	return true, nil
+}
+
+func directoryExists(path string) (bool, error) {
+	info, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	if !info.IsDir() {
+		return false, errors.New("specified file not directory")
+	}
+	return true, nil
+}
+
+func mkdirAll(path string) error {
+	return os.MkdirAll(path, dirMode)
+}
+
+func writeFile(filepath string, data []byte) error {
+	if err := os.Remove(filepath); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	f, err := os.Create(filepath)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+	if err = f.Chmod(fileMode); err != nil {
+		return err
+	}
+	if _, err = f.Write(data); err != nil {
+		return err
+	}
+	return nil
+}
+
+func readFile(filepath string) ([]byte, error) {
+	return ioutil.ReadFile(filepath)
+}
+
+func serializeBucketConfig(bucketCfg *BucketConfig) ([]byte, error) {
+	return yaml.Marshal(bucketCfg)
```

Comment:
any reason you use yaml for bucket config and gob for blobs ? can you use gob for both ?

Reply:
I could use gob for both of them but I wanted bucketConfig to be easily human readable.

Imagine the case where for some reason someone wants to test creating a custom bucket that was not created during startup. In order to do this they would create a new directory and create a bucketConfig file. I want that bucket config file to be easy to create.

Reply:
Fair enough. I know we use yaml all over the place - but consider using json to yaml for things like this - https://noyaml.com/  -